File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {

        int n = s.size();
        if (n <= 1)
            return n;
        int max = 0;
        int left = 0;
        vector<int> last(128, -1);

        for (int i = 0; i < n; i++) {
            int c = s[i];
            if (last[c] >= left) {
                left = last[c] + 1;
            }
            last[c] = i;
            if ((i - left + 1) > max) {
                max = i - left + 1;
            }
        }
        return max;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """

        len_max = 0
        start_temp = 0
        dp = {}

        for i in range(len(s)):
            idx = s[i]
            if idx in dp and dp[idx] >= start_temp:
                start_temp = dp[idx] + 1
            
            dp[idx] = i
            len_max = max((i - start_temp + 1),len_max)

        return len_max
        